File: clients/Go/model/entity_action.go
package model

import "io"
import . "aicup2020/stream"

type EntityAction struct {
    MoveAction *MoveAction
    BuildAction *BuildAction
    AttackAction *AttackAction
    RepairAction *RepairAction
}
func NewEntityAction(moveAction *MoveAction, buildAction *BuildAction, attackAction *AttackAction, repairAction *RepairAction) EntityAction {
    return EntityAction {
        MoveAction: moveAction,
        BuildAction: buildAction,
        AttackAction: attackAction,
        RepairAction: repairAction,
    }
}
func ReadEntityAction(reader io.Reader) EntityAction {
    result := EntityAction {}
    if ReadBool(reader) {
        var MoveActionValue MoveAction
        MoveActionValue = ReadMoveAction(reader)
        result.MoveAction = &MoveActionValue
    } else {
        result.MoveAction = nil
    }
    if ReadBool(reader) {
        var BuildActionValue BuildAction
        BuildActionValue = ReadBuildAction(reader)
        result.BuildAction = &BuildActionValue
    } else {
        result.BuildAction = nil
    }
    if ReadBool(reader) {
        var AttackActionValue AttackAction
        AttackActionValue = ReadAttackAction(reader)
        result.AttackAction = &AttackActionValue
    } else {
        result.AttackAction = nil
    }
    if ReadBool(reader) {
        var RepairActionValue RepairAction
        RepairActionValue = ReadRepairAction(reader)
        result.RepairAction = &RepairActionValue
    } else {
        result.RepairAction = nil
    }
    return result
}
func (value EntityAction) Write(writer io.Writer) {
    if value.MoveAction == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.MoveAction).Write(writer)
    }
    if value.BuildAction == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.BuildAction).Write(writer)
    }
    if value.AttackAction == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.AttackAction).Write(writer)
    }
    if value.RepairAction == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.RepairAction).Write(writer)
    }
}

File: clients/JavaScript/model/entity-action.js
const MoveAction = require('./move-action');
const BuildAction = require('./build-action');
const AttackAction = require('./attack-action');
const RepairAction = require('./repair-action');
class EntityAction {
    constructor(moveAction, buildAction, attackAction, repairAction) {
        this.moveAction = moveAction;
        this.buildAction = buildAction;
        this.attackAction = attackAction;
        this.repairAction = repairAction;
    }
    static async readFrom(stream) {
        let moveAction;
        if (await stream.readBool()) {
            moveAction = await MoveAction.readFrom(stream);
        } else {
            moveAction = null;
        }
        let buildAction;
        if (await stream.readBool()) {
            buildAction = await BuildAction.readFrom(stream);
        } else {
            buildAction = null;
        }
        let attackAction;
        if (await stream.readBool()) {
            attackAction = await AttackAction.readFrom(stream);
        } else {
            attackAction = null;
        }
        let repairAction;
        if (await stream.readBool()) {
            repairAction = await RepairAction.readFrom(stream);
        } else {
            repairAction = null;
        }
        return new EntityAction(moveAction, buildAction, attackAction, repairAction);
    }
    async writeTo(stream) {
        let moveAction = this.moveAction;
        if (moveAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await moveAction.writeTo(stream);
        }
        let buildAction = this.buildAction;
        if (buildAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await buildAction.writeTo(stream);
        }
        let attackAction = this.attackAction;
        if (attackAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await attackAction.writeTo(stream);
        }
        let repairAction = this.repairAction;
        if (repairAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await repairAction.writeTo(stream);
        }
    }
}
module.exports = EntityAction
